lightning/src/ln/channel.rs

```diff
@@ -3460,76 +3645,105 @@ impl<Signer: Sign> Channel<Signer> {
-		if self.is_outbound() {
-			let max_feerate = fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::Normal);
-			if (msg.fee_satoshis as u64) > max_feerate as u64 * closing_tx_max_weight / 1000 {
-				if let Some((last_feerate, _, _)) = self.last_sent_closing_fee {
-					if max_feerate <= last_feerate {
-						return Err(ChannelError::Close(format!("Unable to come to consensus about closing feerate, remote wanted something higher ({}) than our Normal feerate ({})", last_feerate, max_feerate)));
-					}
+		if let Some(msgs::ClosingSignedFeeRange { min_fee_satoshis, max_fee_satoshis }) = msg.fee_range {
+			if msg.fee_satoshis < min_fee_satoshis || msg.fee_satoshis > max_fee_satoshis {
+				return Err(ChannelError::Close(format!("Peer sent a bogus closing_signed - suggested fee of {} sat was not in their desired range of {} sat - {} sat", msg.fee_satoshis, min_fee_satoshis, max_fee_satoshis)));
+			}
+			if max_fee_satoshis < our_min_fee {
+				return Err(ChannelError::Warn(format!("Unable to come to consensus about closing feerate, remote's max fee ({} sat) was smaller than our min fee ({} sat)", max_fee_satoshis, our_min_fee)));
+			}
+			if min_fee_satoshis > our_max_fee {
+				return Err(ChannelError::Warn(format!("Unable to come to consensus about closing feerate, remote's min fee ({} sat) was greater than our max fee ({} sat)", min_fee_satoshis, our_max_fee)));
+			}
+
+			if !self.is_outbound() {
+				// They have to pay, so pick the highest fee in the overlapping range.
+				debug_assert_eq!(our_max_fee, u64::max_value()); // Note that we always allow any fee
+				propose_new_fee!(cmp::min(max_fee_satoshis, our_max_fee));
+			} else {
+				if msg.fee_satoshis < our_min_fee || msg.fee_satoshis > our_max_fee {
+					return Err(ChannelError::Close(format!("Peer sent a bogus closing_signed - suggested fee of {} sat was not in our desired range of {} sat - {} sat after we informed them of our range.",
+						msg.fee_satoshis, our_min_fee, our_max_fee)));
 				}
-				propose_new_feerate!(max_feerate);
+				// The proposed fee is in our acceptable range, accept it and broadcast!
+				propose_new_fee!(msg.fee_satoshis);
 			}
 		} else {
-			min_feerate = fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::Background);
-		}
-		if (msg.fee_satoshis as u64) < min_feerate as u64 * closing_tx_max_weight / 1000 {
-			if let Some((last_feerate, _, _)) = self.last_sent_closing_fee {
-				if min_feerate >= last_feerate {
-					return Err(ChannelError::Close(format!("Unable to come to consensus about closing feerate, remote wanted something lower ({}) than our Background feerate ({}).", last_feerate, min_feerate)));
+			// Old fee style negotiation. We don't bother to enforce whether they are complying
+			// with the "making progress" requirements, we just comply and hope for the best.
+			if let Some((last_fee, _)) = self.last_sent_closing_fee {
+				if msg.fee_satoshis > last_fee {
+					if msg.fee_satoshis < our_max_fee {
```

Comment:
Should these checks be <=?

Reply:
The `last_sent_closing_fee == msg.fee_satoshis` case is handled above in the `if last_fee == msg.fee_satoshis {` block.